### ISSUE TITLE
Runtime gets stuck

### ISSUE THREAD

**Issue by Matthias247** [tokio-rs/tokio#1768]
## Version

current master branch (27e5b41067d01c0c9fac230c5addb58034201a63)

## Platform

Windows 10

## Description

The following program gets stuck in a release build. After some  amount of time, no more printouts will happen. This indicates that a wakeup got lost. I played a bit around, and discovered the following:
- The biggest difference lies within the value of `NR_ITERATIONS` - the smaller it is the faster it gets stuck. Maybe an issue with tasks joining "too early"?
- If the `task::spawn` is removed, everything looks fine
- Caught the issue in the debugger, but it only tells me one thread is waiting on `thread::park`, and others are waiting on IO completion ports
- The amount of runtime threads does not seem to make a difference. I just adjusted it to 1 to have less noise in the debugger.
- It does not seem to happen with the `current_thread` runtime

```rust
use tokio::sync::Mutex;
use tokio::sync::oneshot::channel;
use std::sync::Arc;

async fn do_iteration() {
    // Changing this to a higher value will prevent it from getting stuck
    const NR_ITERATIONS: usize = 5;
    let m = Arc::new(Mutex::new(()));//IntrusiveMutex::new((), true));
    let (sender, receiver) = channel();

    let m = m.clone();
    tokio::spawn(async move {
        for _ in 0..NR_ITERATIONS {
            let _ = m.lock().await;
        }
        sender.send(());
    });

    receiver.await;
}

fn main() {
    const NR_ITERATIONS: usize = 10000;
    const BLOCK_ON_PER_RT: usize = 3000;

    for i in 0 .. NR_ITERATIONS {
        {
            println!("Starting runtime {}", i);
            let mut builder = tokio::runtime::Builder::new();
            builder.thread_pool().num_threads(1);
            let mut rt = builder.build().unwrap();

            for _ in 0 .. BLOCK_ON_PER_RT {
                rt.block_on(async {
                    do_iteration().await;
                });
            }

            println!("Stopping runtime");
        }
        println!("Stopped runtime");
    }
}
````

**Comment by Matthias247** [comment]
FWIW, I was not able to reproduce this on my Macbook (old, dual-core) so far. So it might be either a windows issue, or it only shows up if something is fast enough.

**Comment by swarmer** [comment]
I'm able to reproduce this on a Linux machine, but so far not on my macbook.
It does run a few times slower on my macbook.

I've added the two missing unwraps, but they didn't panic when this deadlocked.

**Comment by carllerche** [comment]
An attempt to repro w/ loom: https://github.com/tokio-rs/tokio/tree/eliza/sticky-mutex

**Comment by Matthias247** [comment]
I root-caused this a bit deeper to the async block inside `tokio::spawn` not getting executed in some iterations by adding and observing some atomic variables.

With the added variables I can observe that `SPAWN_PER_BLOCK_ON` is always 0: The code inside `tokio::spawn` did not get executed. Therefore `BLOCK_ON_STATE` is always 2 (it waits for the async block inside spawn to complete). `BLOCK_ON_PER_RT_COUNT` has an arbitrary value => The issue doesn not necesarily happen immediately after a `Runtime` had been started.

Whether it might have something to do with `spawn` happening shortly after `block_on` is unclear. The only time we ever do a `spawn` in this example is directly after a `block_on`.


```rust
use tokio::sync::Mutex;
use tokio::sync::oneshot::channel;
use std::sync::Arc;
use std::sync::atomic::{AtomicUsize, Ordering};

static BLOCK_ON_PER_RT_COUNT: AtomicUsize = AtomicUsize::new(0);
static BLOCK_ON_STATE: AtomicUsize = AtomicUsize::new(0);
static SPAWN_PER_BLOCK_ON: AtomicUsize = AtomicUsize::new(0);

async fn do_iteration() {
    BLOCK_ON_STATE.store(1, Ordering::Release);

    // Changing this to a higher value will prevent it from getting stuck
    const NR_ITERATIONS: usize = 5;
    let m = Arc::new(Mutex::new(()));//IntrusiveMutex::new((), true));
    let (sender, receiver) = channel();

    let m = m.clone();
    tokio::spawn(async move {
        SPAWN_PER_BLOCK_ON.fetch_add(1, Ordering::Release);
        for _ in 0..NR_ITERATIONS {
            let _ = m.lock().await;
        }
        sender.send(());
    });

    BLOCK_ON_STATE.store(2, Ordering::Release);
    receiver.await;
    BLOCK_ON_STATE.store(3, Ordering::Release);
}

fn main() {
    const NR_ITERATIONS: usize = 10000;
    const BLOCK_ON_PER_RT: usize = 3000;

    for i in 0 .. NR_ITERATIONS {
        {
            BLOCK_ON_PER_RT_COUNT.store(0, Ordering::Release);
            println!("Starting runtime {}", i);
            let mut builder = tokio::runtime::Builder::new();
            builder.thread_pool().num_threads(1);
            let mut rt = builder.build().unwrap();

            for _ in 0 .. BLOCK_ON_PER_RT {
                BLOCK_ON_STATE.store(0, Ordering::Release);
                SPAWN_PER_BLOCK_ON.store(0, Ordering::Release);
                BLOCK_ON_PER_RT_COUNT.fetch_add(1, Ordering::Release);

                rt.block_on(async {
                    do_iteration().await;
                });
            }

            println!("Stopping runtime");
        }
        println!("Stopped runtime");
    }
}
```

**Comment by hawkw** [comment]
> An attempt to repro w/ loom: https://github.com/tokio-rs/tokio/tree/eliza/sticky-mutex

for future readers: it should be noted that this does _not_ actually repro the bug

**Comment by carllerche** [comment]
I have not reproduced this bug yet

**Comment by swarmer** [comment]
I just did some more debugging and maybe I understand what's going on?
First of all, I found out that this is *much* easier to reproduce when setting up 2 threads instead of one, so this is how I was testing it.
Basically, `tokio/src/runtime/thread_pool/idle.rs:worker_to_notify` checks if it needs to notify any workers, sees that one is searching and doesn't notify anything. However *just before this* the worker that was searching tried to find something to steal at `tokio/src/runtime/thread_pool/worker.rs:process_available_work`, didn't find anything, and parked itself.
This is the first time I'm digging into any of this code, so it's quite probable that there's more to the story and there is a reason why this should work anyway, that I don't know about.

**Comment by hawkw** [comment]
@swarmer do you have a more consistent repro with two threads that you can share?

**Comment by swarmer** [comment]
@hawkw I'm just running the first example @Matthias247 provided, but with 1 replaced by 2, and when I reproduced this on latest master, I renamed the builder method according to the latest changes.

**Comment by Matthias247** [comment]
One observation we had yesterday was that this is actually not related to anything inside the subtask, and that just spawning and joining the task is sufficient. Therefore the xample can be reduced to:

```rust

async fn do_iteration() {
    let jh = tokio::spawn(async move {        
    });

    jh.await;
}

fn main() {
    const NR_ITERATIONS: usize = 10000;
    const BLOCK_ON_PER_RT: usize = 3000;

    for i in 0 .. NR_ITERATIONS {
        {
            println!("Starting runtime {}", i);
            let mut builder = tokio::runtime::Builder::new();
            builder.threaded_scheduler().num_threads(2);
            let mut rt = builder.build().unwrap();

            for _ in 0 .. BLOCK_ON_PER_RT {
                rt.block_on(async {
                    do_iteration().await;
                });
            }

            println!("Stopping runtime");
        }
        println!("Stopped runtime");
    }
}


````

**Comment by carllerche** [comment]
@swarmer thanks for the debugging. I am on mobile atm so have limited ability to dig in. 

You are correct regarding the searching hazard. in theory, when the last searching worker exits the searching state before parking, it checks the queues: https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/thread_pool/worker.rs#L547 this should prevent the case you are mentioning.

**Comment by swarmer** [comment]
@carllerche I just found this code and put a log there; and this the end of logs I got:
```
searching: 0, unparked: 0, num_workers: 2
searching: 0, unparked: 0, num_workers: 2
searching: 0, unparked: 1, num_workers: 2
searching: 0, unparked: 1, num_workers: 2
searching work <these two logs are about stealing>
work not found
final sweep empty
searching: 1, unparked: 1, num_workers: 2 <such logs are only printed from the spawning thread, I think>
<it deadlocks here>
```

So apparently this sweep doesn't find anything. I don't have an explanation why.

**Comment by carllerche** [comment]
So you are saying that:

### Injector

- Pushes task
- Sees searching == 1

### Worker

- Decrements searching
- Does final sweep, sees no task.

**Comment by swarmer** [comment]
The exact order/interleaving of events is in question, but basically yes, that's how I understand what I'm seeing. I now want to try to dig into queue stuff more.

**Comment by Matthias247** [comment]
With the `let state = State(self.state.fetch_add(0, SeqCst));` change from #1788 I can not observe the issue anymore in any of the tests that I had been trying so far.

**Comment by hidva** [comment]
Rust uses the C++ memory model, and the C++ standard stipulates:
> An implementation should ensure that the last value (in modification order) assigned by an atomic or synchronization operation will become visible to all other threads in a finite period of time. Implementations should make atomic stores visible to atomic loads within a reasonable amount of time.
> 
> Atomic read-modify-write operations shall always read the last value (in the modification order) written before the write associated with the read-modify-write operation.

This means that load is not guaranteed to see the latest value, but RMW has this guarantee.
